Fix and Enhance Usergroup ACL Permissions Grids and Filtering

smg6511 opened the pull request:
### What does it do?
Made a number of changes and additions in the processors and grid JS files to correct issues with the grids’ grouping functionality and more accurately build the grid filters depending on the data shown. More specifically:

1. Replaced the column model renderers (used to render links on column data) with a `templatecolumn` specification (generated by a new method in the grids base class), as using renderers on grouping grids interferes with the grouping functionality (because a renderer is an interceptor method in Ext).
2. Moved menu specifications (grid context menus) out of processors and into the grid js for all sections (context was already this way). The primary reason being that some events (such as `afterRemoveRow`) were not firing when set up this way; monitoring these events is necessary for maintaining accurate filter options when making changes to a grid's data. Also, there's no reason to specify js in php unless the item in question will change on the fly and needs setup data from the processor.
3. Grid view, including the bottom toolbar, now correctly updates when a row is deleted.
4. Created new methods to update dependent filters (when two or more filters are present) based upon the selected value of any given filter.
5. Fixed some filters that were missing options due to an incorrect specification of their `baseParams.group` property.

### Why is it needed?
Grouping in these grids was not working correctly. Also, providing filter options where no relevant data is available in a given grid is a UX problem that needs to be fixed not only here, but across all grids with filters. 

### How to test
1. Ensure modx and browser caches are fully cleared.
2. Create a number of policies in each area (if you don't already have some set up) and play with the grouping and filtering of items in each grid. (These are the five Access Permissions grids — Context, Resource Groups, Element Categories, Media Sources, and Namespaces.) Verify that filters only show relevant options depending on the grid data and other filters you've made selections on.
3. Verify that other operations (editing and deleting) affect filter options as expected.

### Related issue(s)/PR(s)
Resolves #16239.

### Special Note
Because these updates involve a lot of changes, I didn't want to execute them for _all_ relevant grids at once. This PR completes a logical section and, depending on any requested changes, other PRs will follow what I've done here for other grids.

### Before and After Examples
Currently, in MODx 3.0.3-dev, note each of the following problems (in order of appearance in the video):

1. Each ACL area grid does not group properly; where an area's rows should be grouped under a common heading, they are not. To see this clearly, pause each time I click on each area and note where grouping should be occurring (for example, in Contexts, the three assigned the _Editors - 1000_ Role should be grouped under one row heading for that Role).
2. [0:19] Next, notice how the grid filters show irrelevant options where they should only show those present in the grid data. The _Filter by Context_ combo shows all Contexts instead of those present (dupctx1, dupctx2, dev2, and www2). The same goes for the _Filter by Policy_ combo. [0:42] Additionally, the Filter by Policy combo is mis-configured in places; here the problem in the Contexts area is illustrated. When adding a new Context, you are given the full Policies list in the Create window, but the filter combo shows an incomplete subset. [1:02] Lastly, when attempting to create additional policies for the same Context, an error occurs due to an incorrect validation for a unique context key. 
3. [1:32] When deleting a row, see how the count is not updated in the lower right grid status area.
4. [1:50] Next, when filtering, a choice in one combo should update the others to show only entries relevant to the current state of the query. They do not currently react to changes as one would expect.

https://user-images.githubusercontent.com/689075/209720912-944ca7bb-2d91-4aae-9bdd-f0b5b7d71b18.mov

Now, note how each problem is fixed (action taken in same order as above):

https://user-images.githubusercontent.com/689075/209720965-f3b53940-7554-4db1-a695-86bc2b25c8b7.mov

Commit:
Updates to ACL permissions grids and filtering

Commit:
Updates to ACL permissions grids and filtering

JoshuaLuckers commented on core/src/Revolution/Processors/Context/GetList.php:
```diff
@@ -40,6 +42,9 @@ class GetList extends GetListProcessor
     public $canCreate = false;
 
+    /** @param boolean $isGridFilter Indicates the target of this list data is a filter field */
+    private $isGridFilter;
```
```suggestion
    protected $isGridFilter = false;
```

I think it's better to make this variable protected instead of private, especially since people can extend this class and the variable is used in public methods. But if you have a good argument to make them private, let me know!

smg6511 replied:
Good point, I wasn't considering the possible extension of these GetList classes. I'll change that throughout. 
UPDATE: Change done for all 6 files.

JoshuaLuckers commented on core/src/Revolution/Processors/Context/GetList.php:
```diff
@@ -79,7 +84,28 @@ public function prepareQueryBeforeCount(xPDOQuery $c)
             ]);
         }
-
+        /*
+            When this class is used to fetch data for a grid filter's store (combo),
+            limit results to only those contexts present in the current grid.
+        */
+        if ($this->isGridFilter) {
+            if ($userGroup = $this->getProperty('usergroup', false)) {
+                $c->innerJoin(
+                    modAccessContext::class,
+                    'modAccessContext',
+                    [
+                        '`modAccessContext`.`target` = `modContext`.`key`',
+                        '`modAccessContext`.`principal` = ' . (int)$userGroup,
+                        '`modAccessContext`.`principal_class` = ' . $this->modx->quote(modUserGroup::class)
+                    ]
+                );
+                if ($policy = $this->getProperty('policy', false)) {
+                    $c->where([
+                        '`modAccessContext`.`policy`' => (int)$policy
+                    ]);
+                }
+            }
+        }
```
Have you considered to use an early return of `$c` instead of using a huge if statement block?

smg6511 replied:
There have been other PRs I've worked on where an early return of X was suggested, helping me to learn the benefit of doing so in certain situations. I don't know that it really helps here. To me, doing what I think you're alluding to ...
```
if (!$this->isGridFilter) {
    return $c;
}
if ($userGroup = ...) {
[...]
return $c;
}

```
adds another couple lines and puts that return in a place you might not at first spot when scanning the code. Since an early return doesn't short circuit the parsing of any code, I don't see the benefit. That said, if you feel strongly about it (or obviously if I'm just missing something) I certainly don't mind changing it.

JoshuaLuckers replied:
I don’t have a strong opinion on this, I was just wondering if you did consider the early return and you did.

JoshuaLuckers commented on core/src/Revolution/Processors/Element/Category/GetList.php:
```diff
@@ -33,19 +35,23 @@ class GetList extends GetListProcessor
     public $defaultSortField = 'category';
 
+    /** @param boolean $isGridFilter Indicates the target of this list data is a filter field */
+    private $isGridFilter;
```
```suggestion
    protected $isGridFilter = false;
```

JoshuaLuckers commented on core/src/Revolution/Processors/Security/Access/Policy/GetList.php:
```diff
@@ -35,6 +42,9 @@ class GetList extends GetListProcessor
     public $languageTopics = ['policy', 'en:policy'];
 
+    /** @param boolean $isGridFilter Indicates the target of this list data is a filter field */
+    private $isGridFilter;
```
```suggestion
    protected $isGridFilter = false;
```

JoshuaLuckers commented on core/src/Revolution/Processors/Security/Access/Policy/GetList.php:
```diff
@@ -57,20 +68,101 @@ public function initialize()
+                        if (!empty($category)) {
+                            $c->where([
+                                '`modAccessCategory`.`target`' => (int)$category
+                            ]);
+                        }
+                        break;
+                    case 'MODx.grid.UserGroupSource':
+                        $joinClass = modAccessMediaSource::class;
+                        $joinAlias = 'modAccessMediaSource';
+                        $source = $this->getProperty('source', '');
+                        if (!empty($source)) {
+                            $c->where([
+                                '`modAccessMediaSource`.`target`' => (int)$source
+                            ]);
+                        }
+                        break;
+                    case 'MODx.grid.UserGroupNamespace':
+                        $joinClass = modAccessNamespace::class;
+                        $joinAlias = 'modAccessNamespace';
+                        // Note that namespace pk is a string ('name' is the dataIndex), not an int id
+                        $namespace = $this->getProperty('namespace', '');
+                        if (!empty($namespace)) {
+                            $c->where([
+                                '`modAccessNamespace`.`target`' => $namespace
+                            ]);
+                        }
+                        break;
+                }
+                $c->innerJoin(
+                    $joinClass,
+                    $joinAlias,
+                    [
+                        '`' . $joinAlias . '`.`policy` = `modAccessPolicy`.`id`',
+                        '`' . $joinAlias . '`.`principal` = ' . $userGroup,
+                        '`' . $joinAlias . '`.`principal_class` = ' . $this->modx->quote(modUserGroup::class)
+                    ]
+                );
+                $c->groupby('`modAccessPolicy`.`id`');
+            }
+        }
```
Have you considered to use an early return of `$c` instead of using a huge if statement block?

JoshuaLuckers commented on core/src/Revolution/Processors/Security/ResourceGroup/GetList.php:
```diff
@@ -28,6 +31,48 @@ class GetList extends GetListProcessor
     public $permission = 'resourcegroup_view';
 
+    /** @param boolean $isGridFilter Indicates the target of this list data is a filter field */
+    private $isGridFilter;
```
```suggestion
    protected $isGridFilter = false;
```

JoshuaLuckers commented on core/src/Revolution/Processors/Security/ResourceGroup/GetList.php:
```diff
@@ -28,6 +31,48 @@ class GetList extends GetListProcessor
+     */
+    public function prepareQueryBeforeCount(xPDOQuery $c)
+    {
+        /*
+            When this class is used to fetch data for a grid filter's store (combo),
+            limit results to only those resource groups present in the current grid.
+        */
+        if ($this->isGridFilter) {
+            if ($userGroup = $this->getProperty('usergroup', false)) {
+                $c->innerJoin(
+                    modAccessResourceGroup::class,
+                    'modAccessResourceGroup',
+                    [
+                        '`modAccessResourceGroup`.`target` = `modResourceGroup`.`id`',
+                        '`modAccessResourceGroup`.`principal` = ' . (int)$userGroup,
+                        '`modAccessResourceGroup`.`principal_class` = ' . $this->modx->quote(modUserGroup::class)
+                    ]
+                );
+            }
+            if ($policy = $this->getProperty('policy', false)) {
+                $c->where([
+                    '`modAccessResourceGroup`.`policy`' => (int)$policy
+                ]);
+            }
+        }
+        return $c;
```
Have you considered to use an early return of `$c` instead of using a huge if statement block?

JoshuaLuckers commented on core/src/Revolution/Processors/Source/GetList.php:
```diff
@@ -30,6 +33,9 @@ class GetList extends GetListProcessor
     public $permission = 'source_view';
 
+    /** @param boolean $isGridFilter Indicates the target of this list data is a filter field */
+    private $isGridFilter;
```
```suggestion
    protected $isGridFilter = false;
```

JoshuaLuckers commented on core/src/Revolution/Processors/Workspace/PackageNamespace/GetList.php:
```diff
@@ -30,6 +33,9 @@ class GetList extends GetListProcessor
     public $permission = 'namespaces';
 
+    /** @param boolean $isGridFilter Indicates the target of this list data is a filter field */
+    private $isGridFilter;
```
```suggestion
    protected $isGridFilter = false;
```

JoshuaLuckers reviewed (commented):
This is a huge PR with only 1 commit making it pretty hard for me to review at once. 
I've only reviewed a part of the code for now (without testing the functionality).

Commit:
Update visibility of $isGridFilter

Commit:
Update visibility of $isGridFilter

Commit:
Followup on previous commit - set default value for $isGridFilter

Commit:
Followup on previous commit - set default value for $isGridFilter

Commit:
Update visibility/default value of $isGridFilter

Forgot to update these in the previous commits!

Commit:
Update visibility/default value of $isGridFilter

Forgot to update these in the previous commits!

smg6511 commented:
@JoshuaLuckers - Thanks for beginning to take a look at this! I know it's relatively big, but in the past I've gotten caught between "it's covering too much" and "why did you only apply it to this one little area" (paraphrasing, and not referring to any particular reviewer). I was hoping this would be a happy medium ;-)

I think once you get a closer look you'll see the repeating pattern of changes for the 5 grids, making it a little easier to assess. As with most of my bigger PRs, it seeks to centralize shared routines in the base classes to try to get away from all the needless duplication of similar and often times identical code across related classes.

Also note that I've resisted including other fixes I've identified for these grids, and will have a couple forthcoming PRs that make other incremental improvements that build on the ones here.

JoshuaLuckers commented:
> @JoshuaLuckers - I know it's relatively big, but in the past I've gotten caught between "it's covering too much" and "why did you only apply it to this one little area" (paraphrasing, and not referring to any particular reviewer). I was hoping this would be a happy medium ;-)

It sure is a happy medium, but in this case it would be helpful if the PR would consist of multiple commits. Especially since you described 5 different things this PR does.

The amount of changes we have to review would stay the same but reviewing small changes per commit is much easier (at least for me). We could also ignore certain ch ages easier (like code style changes). It also gives more context to certain code changes and it would make it easier to revert only a portion of the changes when needed instead of everything.

> Also note that I've resisted including other fixes I've identified for these grids

🙈❤️


For me it's fine to leave it as is and so we can go back on topic ;)

mishanthrop commented on core/src/Revolution/Processors/Context/GetList.php:
```diff
@@ -79,7 +84,28 @@ public function prepareQueryBeforeCount(xPDOQuery $c)
+                    'modAccessContext',
+                    [
+                        '`modAccessContext`.`target` = `modContext`.`key`',
+                        '`modAccessContext`.`principal` = ' . (int)$userGroup,
```
For what cases is explicit type casting here?

smg6511 replied:
Although the usergroup property _should_ only be exposed internally, casting it to int ensures only a safe value will be sent in the query. Note that this format of the join's conditions (a non-associative array of strings) does not get sanitized.

JoshuaLuckers commented:
I tried to understand how to test it and what I should see.

Would it be possible to make a small video demonstrating it?

smg6511 commented:
@JoshuaLuckers - Yes, I'll put a couple before and after videos to demonstrate the changes. Give me a day or so...

JoshuaLuckers commented:
> @JoshuaLuckers - Yes, I'll put a couple before and after videos to demonstrate the changes. Give me a day or so...

No hurries!

smg6511 commented:
@JoshuaLuckers - Ok, I put a couple videos up in the main PR description.

JoshuaLuckers commented:
> @JoshuaLuckers - Ok, I put a couple videos up in the main PR description.

Thanks a lot, now I know what I should look for when testing haha.

smg6511 commented:
One thing to be aware of: A while after submitting this PR I discovered that performing additional sorting on the grids made the grouping fall apart. I do have that figured out and will submit that fix in a new PR once this one is incorporated.

smg6511 commented:
@Ibochkarev - I see you've been doing some reviews lately and was hoping you might have some time to take a look at this. I've got some follow-up work that depends on this PR, so would love to see it merged sooner than later.

JoshuaLuckers committed:
Add OpenCollective as funding platform in composer.json (#16347)

JoshuaLuckers committed:
Add additional basic tests for modResource (#16342)

Mark-H reviewed (approved):
Giving this the 👍 based on your very detailed test and a detailed look through the code. You've clearly spent a lot of time and effort on this and it fixes a number of issues, so well done and thank you. 

The only comment I have is that it would be good to prepare some [documentation](https://github.com/modxorg/docs) for the new utility methods on `MODx.Grid`. Those can be useful for others, and we'll also need to have that information as core team in the future when we forgot this PR and need to figure out how it worked ;) 

There may be minor issues or edge cases I've overlooked, but we can address those in the future.

JoshuaLuckers committed:
Update legacy class references in `modElementPropertySet::class` `element_class` column (#16337)

This will fix Custom Property Sets not working after upgrade to MODX 3.

modxbot committed:
New Crowdin updates (#16336)

* New translations file.inc.php (French)

* New translations setting.inc.php (French)

* New translations snippet.inc.php (French)

* New translations source.inc.php (French)

* New translations chunk.inc.php (Portuguese)

JoshuaLuckers committed:
Fix issues with TV's Media Sources after upgrading to MODX 3 (#16338)

* Update legacy class references in `modMediaSourceElement::class` `object_class` column

This will fix an "PHP Fatal error: Cannot declare class modTemplateVarInputRenderText, because the name is already in use" after upgrade to MODX 3.

* Change @property to @var in upgrade script

Co-authored-by: Jason Coward <jason@opengeek.com>

smg6511 commented:
@Mark-H - Ok, I'll put documentation on my near-term agenda. Thanks for reviewing!

Commit:
Grid filtering via URL parameters -- Usergroup ACL (#16089)

* Add remembering filter properties of template TVs grid in browser url

* Fix and improve GetAreas

Fix issue with a non-named area being blank in the list. Also fine-tuned the building of the list array so that duplicate areas show only one list entry along with the correct count of setting entries. Lastly, fixed the sorting of the list to make it based on the label and not the actual value.

* Required method additions for new persistence feature

* Application of persistence feature to ACL User Group panel

* A few pre-submission tweaks

* grunt build

Co-authored-by: Ruslan-Aleev <aleev.insmile@gmail.com>
Co-authored-by: Jason Coward <jason@opengeek.com>

opengeek commented:
Ok, so this has a ton of conflicts after merging #16089. Can these be resolved today so we can include this in 3.0.3 tomorrow?

Commit:
Permissions Combo Fix (#16273)

* Fix getList processor query

* Monitor for delete in combo listener

* Update GetList.php

Make new query variable more descriptive

JoshuaLuckers committed:
Fix TypeError exceptions when restoring deleted resources (#16303)

* Fix TypeError exceptions when restoring deleted resources

* Fix code style

Commit:
Fix Validation Error When Switching TV Types (#16319)

Commit:
Add new error reporting for grids (#16331)

* Added new error reporting for grids

* Update core/lexicon/en/default.inc.php

Co-authored-by: Joshua Lückers <joshualuckers@me.com>

* grunt build

Co-authored-by: Joshua Lückers <joshualuckers@me.com>
Co-authored-by: Jason Coward <jason@opengeek.com>

Commit:
JavaScript linter: step one (#16321)

* Install and configure eslint -- JavaScript linter

* Move frontend stuff (package.json etc) to the root

* Move frontend stuff (package.json etc) to the root

* javascript linter: temporarly disable some rules

opengeek committed:
Fix cache_resource_clear_partial functionality (#16346)

Resolves #16343 for 3.x branch

smg6511 commented:
@opengeek - Yes, I'll take a look at those in just a bit...

Commit:
Merge branch '3.x-improve-usergroup-permissions-grids-filters' of https://github.com/smg6511/revolution into 3.x-improve-usergroup-permissions-grids-filters

Commit:
Conflict Fixes

Various tweaks to bring changes from this PR in line with the recently-merged #16089

smg6511 commented:
@opengeek - Ok, now everything's working together nicely.

opengeek commented:
For future reference, you need to rebase your commits on the latest, not merge the new ones into them.

opengeek commented:
This is going to have to wait for the next release, as there is no way to sort through this mess like this. Can you please revert the merges and rebase the work from just this PR on the up-to-date 3.x branch? When you have time…

smg6511 commented:
@opengeek - I know about the preference to rebase and actually _tried_ to do just that but ran into issues. The changes and conflicts were so intertwined that it seemed to be too confusing for rebase to deal with. It was weird, I started off in the rebase process and ended up to my surprise with a merge happening. (Note that I use Github Desktop and VS Code.) Anyway, given that background, is it easier if I just pull this apart and resubmit in a new PR?

Mark-H commented:
Don't mind me, [just dropping a useful link for anyone looking to get better at rebases](https://www.youtube.com/watch?v=uI1V7771plw).

smg6511 commented:
@Mark-H - Thanks, I _will_ check that out. I usually don't run into issues and have rebased many of my PRs and in-progress work, but this one was a different story!

Mark-H commented:
Closing in favour of #16355
